### PR TITLE
feat: upgrade to OpenTelemetry metrics SDKs

### PR DESCRIPTION
**Pull request by seemk**
# Description

Switch from SignalFx metrics to OpenTelemetry metrics. This will use the RC of OTel metrics SDK.

## Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

- [x] Tested manually
- [x] Added automated tests

# Checklist:

- [x] Unit tests have been added/updated
- [x] Documentation has been updated
- [ ] Change file has been generated (`npm run change:new`)
- [ ] Delete this branch (after the PR is merged)

### CODE REVIEW

**Review by aurbiztondo-splunk — APPROVED** [review]

File: src/start.ts [hunk 1]
```diff
@@ -24,7 +24,7 @@ interface Options {
   serviceName: string;
   // Signal-specific configuration options:
   metrics: boolean | MetricsOptions;
-  profiling: boolean | ProfilingOptions;
+  profiling: ProfilingOptions;
```
[the remark this answers is not in the document]
Comment (seemk) [reply src/start.ts]: removed it

File: src/start.ts [hunk 2]
```diff
@@ -75,8 +79,6 @@ export const stop = async () => {
   const promises = [];
 
   if (running.metrics) {
-    // not actually a promise, for forwards compatibility
-    promises.push(running.metrics.stopMetrics());
```
[the remark this answers is not in the document]
Comment (seemk) [reply src/start.ts:79]: added one

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@splunk/otel",
-  "version": "1.3.0",
+  "version": "1.3.2-metrics-next.9",
```
Comment (Rauno56) [inline package.json:3]: That should not be merged to main.

File: docs/advanced-config.md [hunk 4]
```diff
@@ -59,20 +59,17 @@ The following config options can be set by passing them as arguments to `startTr
...
-
-#### Additional `startMetrics` config options
-
-- `signalfx`: A JS object with optional `client` and `dimensions` fields. If you have already setup a [SignalFx client](https://github.com/signalfx/signalfx-nodejs) with custom configuration, you can use this for sending instead of creating, configuring a new one. `dimensions` object adds a pre-defined dimension for each datapoint. The format for `dimensions` is `{key: value, ...}`.
-
-   The following is a list of dimensions added by default:
-   - `service`: see [`serviceName`](#tracing) from the tracing section
-   - `metric_source`: `splunk-otel-js`
-   - `node_version`: `process.versions.node`, e.g. `16.10.0`
+| `OTEL_SERVICE_NAME`<br>`serviceName`                            | `unnamed-node-service`  | Stable  | The service name of this Node service.
```
[the remark this answers is not in the document]
Comment (Rauno56) [reply docs/advanced-config.md:69]: No-comment-resolve?

File: package.json [hunk 5]
```diff
@@ -97,29 +96,31 @@
...
-    "@opentelemetry/sdk-trace-node": "~1.2.0",
-    "@opentelemetry/semantic-conventions": "~1.2.0",
+    "@opentelemetry/api-metrics": "^0.30.0",
+    "@opentelemetry/context-async-hooks": "~1.4.0",
+    "@opentelemetry/core": "~1.4.0",
+    "@opentelemetry/exporter-jaeger": "~1.4.0",
+    "@opentelemetry/exporter-trace-otlp-grpc": "^0.30.0",
+    "@opentelemetry/exporter-metrics-otlp-grpc": "^0.30.0",
+    "@opentelemetry/sdk-metrics-base": "^0.30.0",
+    "@opentelemetry/instrumentation": "^0.30.0",
```
[the remark this answers is not in the document]
Comment (Rauno56) [reply package.json]: Another no-comment-resolve. What's up? Am I wrong? Don't think it's significant?

File: src/metrics/console_exporter.ts [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+          const ub = boundaries[boundIdx + 1] ?? Infinity;
+          const count = counts[boundIdx + 1];
+          console.log(`  [${lb}, ${ub}) -> ${count}`);
+        }
+      }
+    }
+  }
+}
+
+export class ConsoleMetricExporter implements PushMetricExporter {
```
[the remark this answers is not in the document]
Comment (Rauno56) [reply src/metrics/console_exporter.ts]: I meant the casing, sorry.

**Comment by theletterf** [comment]
This will require a docs update. Ping me once it's released!

File: docs/metrics.md [hunk 7]
```diff
@@ -1,37 +1,140 @@
...
+const { startMetrics } = require('@splunk/otel');
+const { Resource } = require('@opentelemetry/resources');
+const { metrics } = require('@opentelemetry/api-metrics');
+
+// All fields are optional.
+startMetrics({
+  // Takes preference over OTEL_SERVICE_NAME environment variable
+  serviceName: 'my-service',
+  // The suggested resource is filled in via OTEL_RESOURCE_ATTRIBUTES
+  resourceFactory: (suggestedResource: Resource) => {
```
Comment (Rauno56) [inline docs/metrics.md:26]: This is API that should (eventually) be mirrored to tracing as well.
I'd not force the user to know the `Resource` API to use it and just have it return an object that will be passed and merged to detected Resource.

File: src/version.ts [hunk 8]
```diff
@@ -14,4 +14,4 @@
  * limitations under the License.
  */
 
-export const VERSION = '1.3.0';
+export const VERSION = '1.3.2-metrics-next.9';
```
Comment (Rauno56) [inline src/version.ts:17]: This should not be merged.

**Review by mhennoch — APPROVED** [review]

File: docs/advanced-config.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (seemk) [reply docs/advanced-config.md:69]: They are in the OTel spec though?

File: package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (seemk) [reply package.json]: The peer deps were updated

File: docs/metrics.md [hunk 9]
```diff
@@ -1,37 +1,140 @@
...
+      new PeriodicExportingMetricReader({
+        exportIntervalMillis: 1000,
+        exporter: new OTLPMetricExporter({ url: 'http://localhost:4318' })
+      })
+    ]
+  }
+});
+```
+
+## Choosing temporality aggregation type
```
Comment (theletterf) [inline docs/metrics.md:68]: Please explain what aggregation temporality is and when it should be modified: https://github.com/open-telemetry/opentelemetry-proto/pull/192

File: docs/metrics.md [hunk 10]
```diff
@@ -1,37 +1,140 @@
...
+        })
+      })
+    ]
+  }
+});
+```
+
+## Migrating from SignalFx metrics
+
+The SignalFx client is no longer available.
```
Comment (theletterf) [inline docs/metrics.md:91]: I'd rather say that it's deprecated and copy the EoS note.

File: docs/metrics.md [hunk 11]
```diff
@@ -1,37 +1,140 @@
...
+
+const meter = metrics.getMeter('my-meter');
+meter.createObservableGauge('cpu', result => {
+  result.observe(42);
+});
+const counter = meter.createCounter('clicks');
+counter.add(99);
+```
+
+## Runtime metrics
```
Comment (theletterf) [inline docs/metrics.md:117]: @seemk @Rauno56 I understand memory profiling metrics are still to be named?

**Review by theletterf — CHANGES_REQUESTED** [review]

File: docs/metrics.md [hunk 12]
```diff
@@ -1,37 +1,140 @@
...
+const { OTLPMetricExporter } = require('@opentelemetry/exporter-metrics-otlp-grpc');
+const { AggregationTemporality, PeriodicExportingMetricReader } = require('@opentelemetry/sdk-metrics-base');
+
+startMetrics({
+  serviceName: 'my-service',
+  metricReaderFactory: () => {
+    return [
+      new PeriodicExportingMetricReader({
+        exporter: new OTLPMetricExporter({
+          temporalityPreference: AggregationTemporality.DELTA
```
Comment (theletterf) [inline docs/metrics.md:81]: What types are currently available and what's their effect?

**Review by theletterf — CHANGES_REQUESTED** [review]

File: docs/metrics.md [hunk 13]
```diff
@@ -1,37 +1,140 @@
...
+
+startMetrics({ serviceName: 'my-service' });
+
+const meter = metrics.getMeter('my-meter');
+meter.createObservableGauge('cpu', result => {
+  result.observe(42);
+});
+const counter = meter.createCounter('clicks');
+counter.add(99);
+```
```
Comment (theletterf) [inline docs/metrics.md:115]: This is too terse. Please turn this into steps:

1. Replace the `getSignalFxClient` dependency with `opentelemetry/api-metrics` and initialize metrics collection using `startMetrics()`.
2. Replace calls to `getSignalFxClient()` with `metrics` instances. For example...

**Review by theletterf — CHANGES_REQUESTED** [review]

File: docs/metrics.md (hunk 11 above)
Comment (seemk) [reply docs/metrics.md:117]: Current list of metrics that will be used by memory profiling dashboard:
- `process.runtime.nodejs.memory.heap.total`
- `process.runtime.nodejs.memory.heap.used`
- `process.runtime.nodejs.memory.rss`
- `process.runtime.nodejs.memory.gc.pause`
- `process.runtime.nodejs.memory.gc.count`

Plus there might be another metric in the future that currently has no implementation: allocations / allocation rate. Will be added in a separate PR if it's possible to get the data.

File: docs/metrics.md (hunk 13 above)
Comment (seemk) [reply docs/metrics.md]: The thing is they are completely different APIs and the execution model is different as well. I added an additional explanation, that they are semantically equivalent pieces of code and this is how the translation looks like.

File: docs/metrics.md (hunk 12 above)
Comment (seemk) [reply docs/metrics.md:85]: Added the possible options and a description for both

File: docs/metrics.md (hunk 10 above)
Comment (seemk) [reply docs/metrics.md]: I'm not sure the SignalFx Node.js client is deprecated, but I changed the wording that we removed the dependency on this library.

File: docs/metrics.md (hunk 9 above)
Comment (seemk) [reply docs/metrics.md]: Added a link to the spec and tried to explain the difference between aggregation temporality options.

**Review by theletterf — APPROVED** [review]

**Pull request merged**